service/src/main/java/org/cbioportal/service/util/ProfiledCasesCounter.java

```diff
@@ -58,32 +59,42 @@ public void calculate(List<String> molecularProfileIds, List<String> sampleIds,
                 .map(x -> x.getStudyId() + (countByPatients ? x.getPatientId() : x.getSampleId()))
                 .collect(Collectors.toSet());
 
+        
+        Map<String, Set<String>> genePanelToCases = new HashMap<String, Set<String>>();
+        
         for (AlterationCountByGene alterationCountByGene : alterationCounts) {
             final Set<String> profiledCasesForGene = new HashSet<String>();
             Integer entrezGeneId = alterationCountByGene.getEntrezGeneId();
             List<GenePanel> allPanels = new ArrayList<>();
             
+            Set<String> totalProfiledCases = new HashSet<String>();
+            Set<String> allMatchingGenePanelIds = new HashSet<String>();
+            // different calculations depending on if gene is linked to gene panels 
             if (geneGenePanelMap.containsKey(entrezGeneId)) {
-                geneGenePanelMap.get(entrezGeneId).forEach(genePanel -> {
-                    Set<String> casesWithPanelData = genePanelDataMap
+                // calculate and store cases for each gene panel
+                for (GenePanel genePanel : geneGenePanelMap.get(entrezGeneId)) {
+                    if (!genePanelToCases.containsKey(genePanel.getStableId())) {
+                        Set<String> casesWithPanelData = genePanelDataMap
                             .get(genePanel.getStableId())
                             .stream()
-                            // there can be duplicate patient or sample id, append study id
                             .map(x -> x.getStudyId() + (countByPatients ? x.getPatientId() : x.getSampleId()))
                             .collect(Collectors.toSet());
-                    profiledCasesForGene.addAll(casesWithPanelData);
-                    allPanels.add(genePanel);
-                });
-                
-                profiledCasesForGene.addAll(casesWithoutPanelData);
+                        genePanelToCases.put(genePanel.getStableId(), casesWithPanelData);
+                    }
+                }
+                // for every gene panel associated containing the gene, use the sum of unique cases
+                // as well as cases without panel data
+                for (GenePanel genePanel : geneGenePanelMap.get(entrezGeneId)) {
```

Comment:
I think this loop could be combined with the previous loop (the iteration is the same), although I am not positive. If the three inner lines were appended to the inner lines of the previous loop would the data still be valid?
Looping twice is perhaps not too much overhead, if it is helpful for code clarity.